rhp/v4/client.go

```diff
@@ -0,0 +1,666 @@
+func (c *Client) ReviseContract(ctx context.Context, contract types.V2FileContract, sv SignerVerifier, hp rhpv4.HostPrices, actions []rhpv4.WriteAction) (types.V2FileContract, error) {
+	newRevision, err := c.reviseContract(ctx, contract, sv, hp, actions)
+	if err != nil {
+		return types.V2FileContract{}, fmt.Errorf("RPCReviseContract failed: %w", err)
+	}
+	return newRevision, nil
+}
+
+func (c *Client) reviseContract(ctx context.Context, contract types.V2FileContract, sv SignerVerifier, hp rhpv4.HostPrices, actions []rhpv4.WriteAction) (types.V2FileContract, error) {
+	// run rpc
+	newRevision := contract
+	newRevision.RevisionNumber++
+	err := c.withStream(ctx, func(s *mux.Stream) error {
+		// TODO: create payment revision
+
+		// sign revision
+		sv.SignContract(&newRevision)
+
+		// first roundtrip
+		req := rhpv4.RPCModifySectorsRequest{
+			Prices:  hp,
+			Actions: actions,
+		}
+		var resp rhpv4.RPCModifySectorsResponse
+		if err := rhpv4.WriteRequest(s, &req); err != nil {
+			return err
+		} else if err := rhpv4.ReadResponse(s, &resp); err != nil {
+			return err
+		}
+
+		// second roundtrip
+		req2 := rhpv4.RPCModifySectorsSecondResponse{
+			RenterSignature: newRevision.RenterSignature,
+		}
+		var resp2 rhpv4.RPCModifySectorsThirdResponse
+		if err := rhpv4.WriteResponse(s, &req2); err != nil {
+			return err
+		} else if err := rhpv4.ReadResponse(s, &resp2); err != nil {
+			return err
+		}
+
+		// finalise new revision and verify signature
+		newRevision.HostSignature = resp2.HostSignature
+		if !sv.VerifyHostSignature(newRevision) {
+			return errors.New("invalid host signature")
+		}
+		return nil
+	})
+	return newRevision, err
+}
+
+// SignerVerifier is a minimal interface for signing/verifying contracts and
+// transactions as part of performing RPCs.
+type SignerVerifier interface {
+	io.Closer
+
+	// SignContract signs the given contract, setting the
+	// 'RenterSignature' in the process
+	SignContract(contract *types.V2FileContract)
+
+	// SignInputs signs all inputs within the transaction, setting their
+	// signatures in the process
+	SignInputs(txn *types.V2Transaction)
+
+	// VerifyHostSignature verifies the host signature on the given contract
+	VerifyHostSignature(c types.V2FileContract) bool
+
+	// VerifyRenterSignature verifies the renter signature on the given contract
+	VerifyRenterSignature(c types.V2FileContract) bool
+
+	// VerifyTransaction verifies the whole transaction including the renter,
+	// host and input signatures
+	VerifyTransaction(txn types.V2Transaction) error
+}
+
+// singleAddressSignerVerifier is an implementation of the SignerVerifier which
+// assumes that all inputs can be signed with a single key
+type singleAddressSignerVerifier struct {
+	state       consensus.State
+	contractKey types.PrivateKey
+	walletKey   types.PrivateKey
+}
+
+func (s *singleAddressSignerVerifier) Close() error {
+	clear(s.contractKey)
+	clear(s.walletKey)
+	return nil
+}
```

Comment:
I don't think you want to do this. `types.PrivateKey` is a slice, not an array, so we don't need to worry about it getting copied. (And clearing the slices here will clear them elsewhere too -- definitely not what we want!)